README.md
# SmartTextEditor
 
Project Purpose:
The primary objective was to develop a user-friendly text editor in C++ that not only simplifies text editing but also allows for easy adaptation and modification by programmers. The platform's adaptability ensures that changes can be effortlessly coded and implemented, catering to diverse user needs.

Project Features:
The Smart Text Editor offers a unique user flow, suggesting word completions as users type, compressing text for efficient storage, and enabling secure encoding for confidentiality. The ability to save and load text files seamlessly enhances user convenience, fostering a dynamic and interactive editing experience.

Functionalities:

Typing Text:

Implemented a console-based module for forward typing with vector strings, iterators, and loops.
Word Completion Suggestions:

Developed a system using trie trees or B-trees to provide users with expected word completions from a provided dictionary.
Compression of Text:

Employed Huffman encoding for text compression, storing both the compressed text and the compression tree for recovery.
Saving and Loading Text:

Enabled users to save compressed text and compression tree to a local machine, with an option to reload files for continued editing.
Uncompressing of Text:

Implemented Huffman compression tree loading to decompress the text encoding and recover the original text.
Continue Editing:

Ensured users could seamlessly continue editing, saving, and reloading files multiple times.


History.txt
User: ahmad opened/edited file: ahmad.txt Dec 15 2022 16:27:31
User: ahmad opened/edited file: ahmad.txt.txt Dec 15 2022 16:28:6
User: ka opened/edited file: kak Dec 15 2022 18:1:56
User: nawaz opened/edited file: nawaz Dec 15 2022 18:12:23
User: yahoo checked History Dec 15 2022 18:18:37
User: nawaz opened/edited file: nawaz Dec 15 2022 18:20:32
User: hahaah opened/edited file: haha Dec 15 2022 18:22:2
User: ok opened/edited file: ok Dec 15 2022 18:57:33
User: ok opened/edited file: pk Dec 15 2022 19:0:26
User: k opened/edited file: l Dec 15 2022 20:21:41
User: n opened/edited file: n Dec 15 2022 21:43:8


Vector.h
#pragma once

struct out_of_range
{
};


template<class T> class Vector
{
public:
	/* --- Constructor --- */
	//Default Construtor
	Vector();
	explicit Vector(int s);
	explicit Vector(int s, T);

	//copy constructor
	Vector(const Vector& arg);
	//copy assignment
	Vector<T>& operator=(const Vector<T>& arg);
	//Destructor
	~Vector();
	// -------------------------------------- -
		// -------------------- - iterator
	class iterator;
	iterator begin();
	const iterator begin() const;
	iterator end();
	const iterator end()const;
	const iterator cbegin() const;
	const iterator cend() const;
	//0-0-00----------------
	//Capacity
	bool empty() const;//const signifies it is a constant function
	//returns size of allocated storage capacity
	size_t capacity() const;

	//requests a change in capacity
	//reserve() will never deccrase the capacity.
	void reserve(int newmalloc);

	//change the Vector's size
	//if the newsize is smaller, the last elements will be lost.
	//has a default value param for custom values when resizing.
	void resize(int newsize, T val = T());
	//returns the size of the Vector (number of elements).
	size_t size() const;
	//returns the maximum number of elements the Vector can hold
	size_t max_size() const;
	//reduces capacity to fit the size
	void shrink_to_fit();
	//---------------------------------
	//Modifiers
	//removes all elements from the Vector
	//capacity is not changed
	void clear();

	//Inserts element at the back
	void push_back(const T& d);
	//removes the last element from the Vector
	void pop_back();
	//-------------
	//---------------------------
	//element access
	//access elements with bounds checking
	T& at(int n);
	//access elements with bounds checking for constant Vectors.
	const T& at(int n) const;
	//acccess elements, no bounds checking
	T& operator[](int i);
	//access elements, no bounds checking
	const T& operator[](int i) const;
	//returns a reference to the first element
	T& front();
	//returns a reference to the first element
	const T& front() const;
	//returns reference to the first element
	T& back();
	//returns a reference to the last element
	const T& back() const;
	//returns a pointer to the array used by Vector
	T* data();
	//returns a pointer to the array used by Vector
	const T* data() const;
	//----------------------------
private:
	size_t _size; //total elements in the Vector
	T* _elements;//pointer to first element of Vector
	size_t _space; //total amount of space utilized by the Vector array
};


template<class T>
class Vector<T>::iterator
{
public:
	iterator(T* p) :_curr(p)
	{}
	//operator overloading
	iterator& operator++()
	{
		_curr++;
		return *this;
	}

	iterator& operator--()
	{
		_curr--;
		return *this;
	}
	T& operator*()
	{
		return *_curr;
	}

	bool operator==(const iterator& b) const
	{
		return *_curr == *b._curr;
	}

	bool operator!=(const iterator& b) const
	{
		return *_curr != *b._curr;
	}
private:
	T* _curr;
};

//Constructor/Destructor
template<class T>
Vector<T>::Vector() : _size(0), _elements(0), _space(0)
{}

template<class T>
inline Vector<T>::Vector(int s)
	: _size(s), _elements(new T[s]), _space(s)
{
	for (int i = 0; i < _size; i++)
	{
		_elements[i] = T();
	}
}

//initialize a vector with size and assigned all it elements with a number
template<class T>
inline Vector<T>::Vector(int s, T n)
	: _size(s), _elements(new T[s]), _space(s)
{
	for (int i = 0; i < _size; i++)
	{
		_elements[i] = (n);
	}
}

//parameterized constructor to initizalie it from another vector
template<class T>
inline Vector<T>::Vector(const Vector& arg)
	:_size(arg._size), _elements(new T[arg._size])
{
	for (int i = 0; i < arg._size; i++)
	{
		_elements[i] = arg._elements[i];
	}
}
//initialize a vector from vector. It removes the prevoius values of the array an the new vector will have the assigned vector values
template<class T>
inline Vector<T>& Vector<T>::operator=(const Vector<T>& a)
{
	if (this == &a) return *this;// self-assignment not work needed
	//current Vector has enough space, os there is no need for new allocation
	if (a._size() <= _space)
	{
		for (int index = 0; index < a._size; ++index)
		{
			_elements[index] = a._elements[index];
			_size = a._size;
			return *this;
		}
	}
	T* p = new T[a._size];
	for (int index = 0; index < a._size; ++index)
	{
		p[index] = a._elements[index];
	}
	delete[] _elements;
	_size = a._size;
	_space = a._size;
	_elements = p;
	return *this;
}

template<class T>
Vector<T>::~Vector()
{
	delete[] _elements;
}

//iterators
//it returns the first element address in the array 

template<class T>
inline typename Vector<T>::iterator Vector<T>::begin()
{
	return Vector<T>::iterator(&_elements[0]);
}


//it returns the first element address in the array but the return type has an additional keyword constant
template<class T>
inline typename const Vector<T>::iterator Vector<T>::begin() const
{
	return Vector<T>::iterator(&_elements[_size]);
}
//it returns the last element address in the array 
template<class T>
inline typename Vector<T>::iterator Vector<T>::end()
{
	return Vector<T>::iterator(&_elements[_size]);
}

//it returns the last element address in the array  but the return type is constant
template<class T>
inline typename const Vector<T>::iterator Vector<T>::end() const
{
	return Vector<T>::iterator(&_elements[_size]);
}

//it returns the first element address in the array  but the return is constant
template<class T>
inline typename const Vector<T>::iterator Vector<T>::cbegin() const
{
	return Vector<T>::iterator(&_elements[0]);
}

//it returns the last element address in the array but the return type is constant
template<class T>
inline typename const Vector<T>::iterator Vector<T>::cend() const
{
	return Vector<T>::iterator(&_elements[_size]);
}

//capacity
//returns true if no elemnet is inserted till now
template<class T>
inline bool Vector<T>::empty() const
{
	return(_size == 0);
}

//returns the capacity which is the value in _space
template<class T>
inline size_t Vector<T>::capacity() const
{
	return _space;
}

template<class T>
inline void Vector<T>::reserve(int newalloc)
{
	//if the new size is already equal or less then the prevois size then no need to resize
	if (newalloc <= _space) return;
	//other wise create new pointer array of data type T
	T* p = new T[newalloc];

	//size the prevoius array elements in it
	for (int i = 0; i < _size; i++)
	{
		p[i] = _elements[i];
	}
	//delete the previous array
	delete[] _elements;
	//finally make teh _elements array the array with big(new size) size
	_elements = p;
	_space = newalloc;//save the new size in _space varaible
}

//function to resize the vector with new size the
template<class T>
inline void Vector<T>::resize(int newsize, T val)
{
	reserve(newsize);
	for (int index = _size; index < newsize; index++)
	{
		_elements[index] = T();
	}
	_size = newsize;
}

//get the current size of the inserted elements in the vector or count of elements in the vector
template<class T>
inline size_t Vector<T>::size() const
{
	return _size;
}

//modifiers
//insert value at the last of the vector
template<class T>
inline void Vector<T>::push_back(const T& d)
{
	if (_space == 0)
		reserve(8);
	else if (_size == _space)
		reserve(2 * _space);
	_elements[_size] = d;
	++_size;
}

//accessors
template<class T>
inline T& Vector<T>::at(int n)
{
	if (n < 0 || _size <= n) throw out_of_range();
	return _elements[n];
}

template<class T>
inline const T& Vector<T>::at(int n)const
{
	if (n < 0 || _size <= n) throw out_of_range();
	return _elements[n];
}


//Operator overloading

template<class T>
inline T& Vector<T>::operator[](int i)
{
	return _elements[i];
}

template<class T>
inline const T& Vector<T>::operator[](int i) const
{
	return _elements[i];
}

template<class T>
inline T& Vector<T>::front()
{
	return _elements[0];
}
template<class T>
inline const T& Vector<T>::front() const
{
	return _elements[0];
}


template<class T>
inline T& Vector<T>::back()
{
	return _elements[_size - 1];
}


template<class T>
inline const T& Vector<T>::back() const
{
	return _elements[_size - 1];
}

template<class T>
inline T* Vector<T>::data()
{
	return _elements;
}

template<class T>
inline const T* Vector<T>::data() const
{
	return _elements;
}

Source.cpp
//#include<iostream>
//#include"Date_Time.h"

#include"Input.h"
//#include"Huffman.h"
//#include"TrieTree.h"
//#include"Vector.h"


//global variable

	struct node1
	{
		char data;
		string brr;
		node1* next;
		node1* previous;
	};
	
	struct NODE {
		char data;
		NODE* LEFT;
		NODE* RIGHT;
		int frequency;
	};

	struct node :public NODE {
		char  data;
		node* next;
		int frequency;
		NODE* obj;
	};

	string name_of_file;
	node* front;
	node* rear;
	node1* head;//for linklist
	node1* tail;

	node* swap(node * ptr1, node * ptr2)
	{
		node* tmp = ptr2->next;
		ptr2->next = ptr1;
		ptr1->next = tmp;
		return ptr2;
	}

	bool isempty()
	{
		if (head == nullptr)
		{
			return 1;
		}
		else return 0;
	}

	void insert_end(char data, string b1)
	{
		node1* ptr = new node1();
		bool flag;
		flag = isempty();
		if (flag == 1)//checking if head is null then it put head to ptr
		{
			head = ptr;
			tail = ptr;
			ptr->data = data;
			ptr->brr = b1;
			ptr->next = nullptr;
			ptr->previous = nullptr;
		}
		else {
			ptr->brr = b1;
			ptr->data = data;
			node1* copy;
			copy = head;
			while (copy->next != nullptr)//loop runs until it find last node
			{
				copy = copy->next;
			}
			copy->next = ptr;
			ptr->previous = copy;
			tail = ptr;
			ptr->next = nullptr;
		}
	}

	string findingcode(char a)
	{
		node1* ptr = head;
		while (ptr != nullptr)
		{
			if (ptr->data == a)
			{
				return (ptr->brr);
			}
			ptr = ptr->next;
		}

	}

	void makingpriotyqueue()
	{
		int count = 0;
		node** head = &front;
		node* start = front;
		while (start != NULL) { //grab count
			count++;
			start = start->next;
		}
		struct node** h;
		int i, j, swapped;

		for (i = 0; i <= count; i++)
		{

			h = head;
			swapped = 0;

			for (j = 0; j < count - i - 1; j++)
			{

				node* p1 = *h;
				node* p2 = p1->next;

				if (p1->frequency > p2->frequency)
				{

					/* update the link after swapping */
					*h = swap(p1, p2);
					swapped = 1;
				}

				h = &(*h)->next;
			}

			/* break if the loop ended without any swap */
			if (swapped == 0)
				break;
		}
	}

	void enqueue(char a, int freq, bool symbol, NODE * kl)///making proity queueu 
	{
		if (front == nullptr && rear == nullptr)
		{
			node* newnode = new node();
			front = newnode;
			rear = newnode;
			newnode->data = a;
			newnode->frequency = freq;
			newnode->obj = kl;
			newnode->obj->data = a;
			newnode->obj->frequency = freq;

		}
		else
		{
			bool flag = 0;
			if (symbol == 0)
			{

				node* check = front;
				while (check != NULL)
				{
					if (check->data == a)
					{
						check->frequency++;////last rear masla may be
						check->obj->frequency++;
						flag = 1;
						break;
					}
					check = check->next;
				}
			}
			if (flag == 0)
			{
				node* check = front;
				while (check->next != NULL)
				{
					check = check->next;
				}
				node* newnode = new node();
				check->next = newnode;
				newnode->data = a;
				newnode->frequency = freq;
				newnode->obj = kl;
				newnode->obj->data = a;
				newnode->obj->frequency = freq;
				rear = newnode;
			}
		}
	}

	void dequeue()
	{
		if (front == nullptr && rear == nullptr)
		{
			cout << "queue is empty";
		}
		else if (front == rear)
		{
			front = nullptr;
			rear = nullptr;
		}
		else {
			node* temp = front;
			front = front->next;
			delete temp;
		}
	}

	void displayqueue()
	{
		node* temp = front;
		cout << "updated queue is:";
		while (temp != nullptr) {
			cout << temp->data << " " << temp->frequency << endl;
			temp = temp->next;
		}
		cout << endl;
	}

	int isSizeOne()
	{
		int counter = 0;
		node* check = front;
		while (check != NULL)
		{
			counter++;
			check = check->next;
		}
		return counter;

	}

	NODE* extractmin()
	{
		node* check = front;
		NODE* temp = new NODE();
		temp = check->obj;
		temp->data = check->data;
		temp->frequency = check->frequency;
		dequeue();
		return temp;
	}

	NODE* huffmantree() //here
	{
		NODE* LL;
		NODE* RR;
		NODE* TOP;
		while (isSizeOne() != 1)
		{
			LL = extractmin();
			RR = extractmin();
			TOP = new NODE();
			TOP->data = '$';
			TOP->frequency = LL->frequency + RR->frequency;
			TOP->LEFT = LL;
			TOP->RIGHT = RR;
			enqueue('$', TOP->frequency, 1, TOP);//also want to wnqueue node actuall
			makingpriotyqueue();

		}
		return(extractmin());

	}

	int check(int a)
	{
		if (a == 0)
		{
			return 48;
		}
		else if (a == 1)
		{
			return 49;
		}
	}

	string printArr(int brr[], int n)
	{
		string h1;
		int i;
		for (i = 0; i < n; ++i)
		{
			//cout << brr[i];
			h1 += check(brr[i]);
		}
		//	cout << "\n";
		return h1;
	}

	int isLeaf(NODE * rooot)
	{

		return !(rooot->LEFT) && !(rooot->RIGHT);
	}

	void printshfcodes(NODE * fun, int ARR[], int n)
	{
		if (fun->LEFT)
		{

			ARR[n] = 0;
			printshfcodes(fun->LEFT, ARR, n + 1);
		}

		// Assign 1 to right edge and recur
		if (fun->RIGHT) {

			ARR[n] = 1;
			printshfcodes(fun->RIGHT, ARR, n + 1);
		}
		if (isLeaf(fun)) {
			string b10;
			//cout << fun->data << ": ";
			b10 = printArr(ARR, n);
			insert_end(fun->data, b10);
			///enqueue
		}

	}

	void encoding()
	{
		int* ARR = new int[100000000];
		bool flag = 0;
		char c1;
		ifstream f1;

		f1.open(name_of_file + ".txt");/////change filename here
		if (!f1) {
			cout << "file not found:";
			exit(0);
		}
		while (!f1.eof())
		{
			NODE* PTR2 = new NODE();
			f1 >> c1;
			if (f1.eof())
			{
				flag = 1;
				break;
			}
			enqueue(c1, 1, 0, PTR2);
		}
		makingpriotyqueue();
		//displayqueue();
		NODE* HFTREE;
		HFTREE = huffmantree();
		printshfcodes(HFTREE, ARR, 0);
	}

	void tablewrite()
	{
		char a;
		string q;
		ofstream f4;
		f4.open(name_of_file + ".table.txt");
		node1* ptr = head;
		while (ptr != NULL)
		{
			a = ptr->data;
			q = ptr->brr;
			f4 << a;
			f4 << q;
			f4 << endl;
			ptr = ptr->next;
		}
		f4.close();
	}

	void makingcompressedfile()
	{
		string d1;
		bool flag;
		char c1;
		ifstream f1;
		ofstream f3;
		f1.open(name_of_file + ".txt");
		f3.open(name_of_file + ".compress.txt");
		while (!f1.eof())
		{
			f1 >> c1;
			if (f1.eof())
			{
				flag = 1;
				break;
			}
			d1 = findingcode(c1);
			f3 << d1;
		}
		f1.close();
		f3.close();
		tablewrite();
	}

	char checking_ifpresentinlist(string w)
	{
		bool flag = 0;
		char q;
		node1* check = head;
		while (check != nullptr)
		{
			if (check->brr == w)
			{
				flag = 1;
				q = check->data;
				break;

			}
			check = check->next;
		}
		if (flag == 0)
		{
			return 48;
		}
		else return q;
	}

	void convetingcompress_to_real()
	{
		char fun;
		char z;
		string q;
		ifstream f2;
		ofstream f3(name_of_file + "_2.txt");
		f2.open(name_of_file + ".compress.txt");
		if (!f2)
		{
			cout << "not opened";
		}
		while (!f2.eof())
		{
			f2 >> z;
			if (f2.eof())
				break;
			q += z;
			fun = checking_ifpresentinlist(q);
			if (fun != 48)
			{
				f3 << fun;
				q.clear();
			}
		}
	}

	void decoding1()
	{
		char a;
		string b;

		int input;
		/*cout << "enter the file on which your compressed data is stored:";
		cin >> input;*///check how do it?
		ifstream f1;
		f1.open(name_of_file + ".table.txt");
		if (!f1)							//if file does not open
		{
			cout << "File could not be opened\n";
			exit(0);
		}
		else {
			while (!f1.eof())
			{
				f1 >> a;
				f1 >> b;
				if (f1.eof())
				{
					break;
				}
				insert_end(a, b);

			}
		}
		//display10();
		convetingcompress_to_real();

	}



int main()
{
	fstream my_file, your_file;
	/*cout << "Hello world\n";
	cout << date_time.GenerateTime() << endl;*/
	Date_Time date_time;
	Input input;
	TrieTree tree;
	//ask username and the file name that we may want use for writing
	input.AskDetails();

	input.InsertDictionary(tree);
	string option;
	
		cout << "1. Open notepad\n";
		cout << "2. Search for file\n";
		cout << "3. Check History\n";
		cout << "4. Clear Screen\n";
		cout << "5. Do HuffMan Compression\n";
		cout << "6. Change your username\n";
		cin >> option;
		cout << "Your entered option is: ";
		switch (option[0])
		{
		case 'n':
			break;

		case 'N':
			break;
		case '1':
			cout << "Launching notepad!\n";
			
			input.LaunchNotepad(tree,my_file,your_file);
			break;
		case '2':
			cout << "Search function called\n";
			input.SearchFile(my_file,your_file);
			break;
		case '3':
			cout << "Beloved History:\n";
			input.SearchHistory(my_file,your_file);
			break;
		case '4':
			system("cls");
			break;
		case '5':
			name_of_file = input.file_name;
			    encoding();
			    makingcompressedfile();
			   decoding1();
			break;
		case '6':
			cout << "Enter the username\n";
			cin >> input.user_name;
			break;
		case '7':
			cout << "Not constructed yet!\n";
			break;
		default:
			cout << "Please enter a valid option";
			break;
		}
	
	return -0;
}

TrieTree.h
#pragma once
#include <iostream>
#include"Vector.h"
#include<queue>
#include<ctime>
#include<cctype>
#include"Date_Time.h"
#include<string>
#include<fstream>
//Global variables
////my_file to write into file and my_file2 is to search for a file and print its data on screen
//disabling un_usual warning
#pragma warning(disable : 4996)

using namespace std;

class TrieNode
{
public:
	//signifies the completion of the word
	bool is_word;
	TrieNode* child[26];
};

class TrieTree {
	TrieNode* root;


public:

	//Default constructor
	TrieNode* GetNewNode();
	TrieTree();
	TrieNode* GetRoot();


	//insert key in the tree
	void Insert(std::string key);

	//additional function to search a particular key
	bool Search(std::string key);

	/*
	helper function takes 3 parameters
	1) root of subtrie of Trie Tree (or subtree)
	2) processing key (key which is going to be deleted). key is string
	3) index -> which character is processing or how much depth

	It returns the root of the subtrie
	*/
	//associated with the delete function
	TrieNode* Helper(TrieNode* root, std::string key, int index);

	TrieNode* DeleteKey(std::string key);
	//find all words related to a prefix or which a key prefix
	//function associated with the auto_complete function. Helper function of auto_complete
	void find_all_words(TrieNode* root, std::string key, Vector<std::string>& all_words);
	std::string auto_complete(TrieNode* root, std::string key);

};

Input.cpp
#include"Input.h"

//Taking obj.GetRoot() node of the Trie Tree and fstream object to write and read data into the user file
void Input::LaunchNotepad(TrieTree& obj,fstream& my_file,fstream& my_file2)
{
   // my_file.flush();
    my_file.open(file_name+".txt", std::ios::app);
    char* ch_array;
    if (!my_file.is_open())
        std::cout << "File is not open\n";
    else
        std::cout << "File is opened/created. Now you can type into the file\n";
    //created a variable of string and str
    std::string string, str;
    //creating a size variable and initializing it with 0
    int _size = 0;
    //creating a character variable initiliazing it with NULL value
    char ch = NULL;
    bool condition = true;
    Date_Time date_time;
    while (condition)
    {
        //As soon as a key from keyboard is pressed then this if will be executed
        if (_kbhit())//taking input form the keyboard
        {
            switch (_getch())//taking the pressed key ascii from getch() as it is stored in getch() when ever kbhit function is used for taking the input keystroke
            {
            case 'a':
                std::cout << 'a';
                string += 'a';
                my_file << 'a';
                break;

            case 'b':
                std::cout << 'b';
                string += 'b';
                my_file << 'b';
                break;

            case 'c':
                std::cout << 'c';
                string += 'c';
                my_file << 'c';
                break;

            case 'd':
                std::cout << 'd';
                string += 'd';
                my_file << 'd';
                break;

            case ('e'):
                std::cout << 'e';
                string += 'e';
                my_file << 'e';
                break;

            case  'f':
                std::cout << 'f';
                string += 'f';
                my_file << 'f';
                break;

            case ('g'):
                std::cout << 'g';
                string += 'g';
                my_file << 'g';

                break;
            case  'h':
                std::cout << 'h';
                string += 'h';
                my_file << 'h';
                break;

            case ('i'):
                std::cout << 'i';
                string += 'i';
                my_file << 'i';

                break;
            case  'j':
                std::cout << 'j';
                string += 'j';
                my_file << 'j';
                break;

            case ('k'):
                std::cout << 'k';
                string += 'k';
                my_file << 'k';
                break;

            case  'l':
                std::cout << 'l';
                string += 'l';
                my_file << 'l';
                break;

            case ('m'):
                std::cout << 'm';
                string += 'm';
                my_file << 'm';
                break;

            case  'n':
                std::cout << 'n';
                string += 'n';
                my_file << 'n';
                break;

            case ('o'):
                std::cout << 'o';
                string += 'o';
                my_file << 'o';
                break;

            case  'p':
                std::cout << 'p';
                string += 'p';
                my_file << 'p';
                break;

            case ('q'):
                std::cout << 'q';
                string += 'q';
                my_file << 'q';
                break;

            case  'r':
                std::cout << 'r';
                string += 'r';
                my_file << 'r';
                break;

            case ('s'):
                std::cout << 's';
                string += 's';
                my_file << 's';
                break;

            case  't':
                std::cout << 't';
                string += 't';
                my_file << 't';
                break;

            case ('u'):
                std::cout << 'u';
                string += 'u';
                my_file << 'u';
                break;

            case  'v':
                std::cout << 'v';
                string += 'v';
                my_file << 'v';
                break;

            case ('w'):
                std::cout << 'w';
                string += 'w';
                my_file << 'w';
                break;

            case  'x':
                std::cout << 'x';
                string += 'x';
                my_file << 'x';
                break;

            case ('y'):
                std::cout << 'y';
                string += 'y';
                my_file << 'y';
                break;

            case  'z':
                std::cout << 'z';
                string += 'z';
                my_file << 'z';
                break;

            case 27://when ESC key is pressed
                condition = false;
                break;

            case 13://new line ascii
                std::cout << "\n";
                string = "";//empty the string
                my_file << '\n';
                break;

            case 32://space ascii
                std::cout << " ";
                string = "";//empty the string
                my_file << ' ';
                break;

            case '.':
                std::cout << ".";
                my_file << '.';
                break;

            case ',':
                std::cout << ",";
                my_file << ',';
                break;

            case '!':
                std::cout << "!";
                my_file << '!';
                break;

            case '/':
                std::cout << "/";
                my_file << '/';
                break;

            case 224:
                for (int a = string.length(); a < str.size(); a++)
                {
                    std::cout << str[a];
                    my_file << str[a];
                }
                //replacing the string with the new one
                string = str;
                break;
            }
            str = (obj.auto_complete(obj.GetRoot(), string));
        }
    }
    //Now adding an new line in it then closing the file
    my_file.close();
    cout << "The file is saved with name: " << file_name << "\n";
    my_file2.open("History.txt", std::ios::app);
    //creating in object of time and date class to generate time
    my_file2 << "User: " << user_name << " opened/edited file: " << file_name << " " << date_time.GenerateTime() << "\n";
    my_file2.close();
}

void Input::SearchFile(fstream& my_file,fstream& my_file2)
{
    std::string line;
    Date_Time date_time;
    std::cout << "\n Enter the file name whose content you would like to see!\n";
    std::cin >> line;

    //my_file.flush();
    my_file.open(line + ".txt", std::fstream::in);//file in input mode to get data form the file
    char ch;
    if (my_file.is_open())
    {
        //print character by charcter from file till the end of the file
        while (my_file.eof() == 0)//if the end of file is not reached then again iterate
        {
            my_file.get(ch);
            std::cout << ch;
        }
    }
    else std::cout << "Can't open the file\n";// in case, unable to open the file
    //In case the ofstream object do not work. So its the quick fix for it
    //my_file2.flush();
    //Noting the time and date of file opening in the History file
    my_file2 << "User: " << user_name << " searched file: " << file_name << " " << date_time.GenerateTime() << "\n";

}

void Input::SearchHistory(fstream& my_file,fstream& my_file2)
{
    std::string line;
    Date_Time date_time;
    //my_file.flush();
    my_file.open("History.txt", std::fstream::in);//file in input mode to get data form the file
    char ch;
    if (my_file.is_open())
    {
        //print character by charcter from file till the end of the file
        while (my_file.eof() == 0)//if the end of file is not reached then again iterate
        {
            
            my_file.get(ch);
            std::cout << ch;
        }
    }
    else std::cout << "Can't open the file\n";// in case, unable to open the file
    //In case the ofstream object do not work. So its the quick fix for it
   // my_file2.flush();
    //Noting the time and date of file opening in the History file
    my_file2.open("History.txt", ios::app);
    my_file2 << "User: " << user_name << " checked History " << date_time.GenerateTime() << "\n";

}


void Input::AskDetails()
{
    std::cout << "Enter your name?\n";
    std::getline(std::cin, user_name);
    std::cout << "Enter the name of the file\n";
    std::cin >> file_name;
    //clearing everyting especially the white space from the cin operator
    std::cin.ignore();
    //adding txt extension to the file name
    
}


void Input::InsertDictionary(TrieTree obj)
{
    std::ifstream dictionary_file("dictionary.txt");
    int index = 0;
    if (dictionary_file.is_open())
    {
        //getting each every character(except the newline ascii) from the file saving it into string variable line
        while (getline(dictionary_file, line))//it will saving each line in the string variable until the last line from the file is not taken/met
        {
            //Inserting the file word string in the tree
            obj.Insert(line);
        }
    }
    else// in case the program is not able to open the file
        std::cout << "Unable to open the file\n";

}


//
//void Input::HuffManCompression() {
//    
//    cout << file_name << endl;
//    huffman.setFileName(file_name);
//    huffman.encoding();
//    huffman.makingcompressedfile();
//    huffman.decoding1();
//}
